Refactored `wp-env` Config Parsing

ObliviousHarmony opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This pull request refactors the existing `.wp-env.json` config file parsing in order to more cleanly support root properties. It aimed to increase composition and testability while maintaining full compatibility.

## Why?
While working on https://github.com/WordPress/gutenberg/pull/49996 we decided that it would be best if the `postInstall` script existed at the configuration root. When I looked into implementing that, however, it proved to be troublesome due to the way config parsing was designed. It was explicitly created to cascade the config file almost immediately, and as a result, it was not easy to introduce root-level options and validation in a way that would not cascade.

<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

## How?
We delay the cascade of options from the root to the environments until after validation and parsing have completed. This lets us interact with the config object while the root and environments are both present but separate. As a result we will be able to support root-level configuration options without any additional changes.

I was worried about breaking compatibility, so, this pull request also features an exhaustive test suite update. All new functionality is more or less completely covered for both happy path and failure states. This was a great exercise and caught a number of bugs in the process.

## Testing Instructions

1. Run `npm run env -- start --update` and make sure that both the development and tests environments work.
2. Run `npm run test:unit packages/env` and make sure that they pass.
3. Create a `.wp-env.override.json` file and make sure that it overrides as expected. Config and mappings should be merged while plugins and themes are overridden.

ObliviousHarmony committed:
Refactored Configuration Parsing

This commit changes the way that configuration parsing works in order
to better facilitate the creation of root-only configuration options.
This does, however, leave `wp-env` in a non-functional state. Now
that the config parsing itself is done, I wanted to commit it. I almost
accidentally deleted this once today and that was too many times!

ObliviousHarmony committed:
Added Config Validation Tests

ObliviousHarmony committed:
Expanded Test Suite

This commit adds basically all of the other tests from
`tests/config.js` in their respective test files. This covers the
last of the tests and the config parsing should now have the
same or better level of test coverage.

ObliviousHarmony committed:
Added Unique Port Validation

ObliviousHarmony committed:
Fixed Tests

ObliviousHarmony committed:
Fixed Test Promises & Setup

ObliviousHarmony committed:
Clarified Default Type Inference

ObliviousHarmony committed:
Refactored getCacheDirectory

In order to create the `WPConfig` object we will switch to
composing the output from config parsing into a new class.
This lets us avoid having to rewrite all of the tests to get
a different output and continue with the trend of
composing together the parsing logic.

ObliviousHarmony committed:
Replaced Existing Config Loading

This commit finishes the refactor and hooks up the loading
of config files using the new functionality.

ObliviousHarmony committed:
Added Config Parsing Integration Test

ObliviousHarmony committed:
Finished Integration Test

ObliviousHarmony committed:
Fixed Parse Config Snapshot Test

noahtallen commented on packages/env/lib/config/load-config.js:
```diff
@@ -0,0 +1,89 @@
+		),
+		configDirectoryPath,
+		workDirectoryPath: cacheDirectoryPath,
+		detectedLocalConfig: await hasLocalConfig( [
```
Possibly worth adding this to the config object itself instead of doing a second check for the local config here

ObliviousHarmony replied:
I thought about doing that, but, I wanted to keep the simplicity of having the config only contain configuration options. Since this check is only used by `start.js`, I would probably rather just move `hasLocalConfig` there, to be honest.

noahtallen commented on packages/env/lib/config/parse-config.js:
```diff
@@ -3,206 +3,421 @@
+
+	// Merge all of our configs so that we have a complete object
+	// containing the desired options in order of precedence.
+	return mergeConfigs(
```
Nice, I like how clear this is 👍

noahtallen commented on packages/env/lib/config/parse-config.js:
```diff
@@ -3,206 +3,421 @@
+) {
+	// Our default config should try to infer what type of project
+	// this is in order to automatically map the current directory.
+	const type = shouldInferType
```
```suggestion
	const detectedType = shouldInferType
```

just to make it more clear that `type` might not apply in every situation

ObliviousHarmony replied:
Agreed

noahtallen commented on packages/env/lib/config/parse-config.js:
```diff
@@ -3,206 +3,421 @@
+ *
+ * @return {WPServiceConfig} An object containing the environment variable overrides.
+ */
+function getEnvironmentVarOverrides( cacheDirectoryPath ) {
```
I noticed that this config object works differently from the other pieces, and I'm wondering if it should also go through the parser/validator? For example, I notice in `getConfigFromEnvironmentVars` that we do the source parsing when reading the environment variables, but for the other configs, we do source parsing during `parseServiceConfig`. 

I wonder if it'd be simpler to just have a single function for this (instead of also having `getConfigFromEnvironmentVars`), and then just run that more barebones config object through `parseRootConfig` (or `parseServiceConfig`, whichever makes sense) and let that handle the extra steps

ObliviousHarmony replied:
This was mostly just a consequence of copy/pasting the original environment variable parsing with minimal changes. One consideration though is that, as-is, we get the nice presentation of error messages from the validation done there. If we used `parseServiceConfig()` we would get the `configFile` and `development.port` style error messages.

I think that, although it's different, I would rather keep it the way it is. It seems better to perform the parsing on environment variables in a very granular way like this instead of trying to pass it through the other parsing logic.

noahtallen replied:
Fair enough 👍

noahtallen commented on packages/env/lib/config/post-process-config.js:
```diff
@@ -0,0 +1,195 @@
+
+	// Set any root-level options we reset back.
+	for ( const option in removedRootOptions ) {
+		config[ option ] = removedRootOptions[ option ];
```
I'm curious about the value of maintaining both the root config and the individual env configs. Do we ever need to access the root config directly internally?

ObliviousHarmony replied:
In cases like this I think it's important for the output structure to match the input structure. You wouldn't expect post-processing on your config object to discard parts of it. Even if we don't really need this right now, I don't really think there's a problem with maintaining it until the consumer decides they don't need it.

> Do we ever need to access the root config directly internally?

No; it's also discarded almost immediately after this in `loadConfig`. Only the `config.env` keys are passed to the rest of the utility. This won't always be the case though. Once we have root properties like `postInstall`, we will want to bubble those up to `loadConfig()` so it can be put into the `WPConfig` object returned there.

noahtallen commented on packages/env/lib/config/post-process-config.js:
```diff
@@ -0,0 +1,195 @@
+	// We are only interested in editing the config options for environment-specific configs.
+	// If we made this change to the root config it would cause problems since they would
+	// be mapped to all environments even though the ports will be different.
+	for ( const env in config.env ) {
```
Since we're accessing `config.env[env]` so much, I might use `for (const [_, env] of Object.entries(config.env))` instead, and then `env` is just the total env object

ObliviousHarmony replied:
Honestly, I kind of like things like this to be a little bit more expressive about what exactly you're editing. I can change it if you really want me to though :smile:

noahtallen replied:
I do personally feel like it's more readable and a bit less noisy with object.entries, but I don't care that much :)

noahtallen commented on packages/env/lib/config/post-process-config.js:
```diff
@@ -0,0 +1,195 @@
+	// overwrite any root-level options if necessary.
+	for ( const env in config.env ) {
+		if ( config.env[ env ].port === undefined ) {
+			throw new ValidationError( `The "${ env }" has an invalid port.` );
```
I think this would come out as "The development has an invalid port" -- might need some clarification :) 

side note -- is it possible to get into this condition at this point?

ObliviousHarmony replied:
You're right, it would. As for the condition itself, no, I don't think so. It was possible in an earlier version of this class, but, I don't think so now. I'd rather leave the extra validation though, just in case :)

noahtallen commented on packages/env/lib/config/post-process-config.js:
```diff
@@ -0,0 +1,195 @@
+
+			if ( environmentPorts[ env ] === environmentPorts[ check ] ) {
+				throw new ValidationError(
+					`The "${ env }" and "${ check }" environments may not have the same port.`
```
Nice improvement 👍

noahtallen commented on packages/env/lib/config/read-raw-config-file.js:
```diff
@@ -13,26 +14,27 @@ const { ValidationError } = require( './validate-config' );
-module.exports = async function readRawConfigFile( name, configPath ) {
+module.exports = async function readRawConfigFile( configPath ) {
 	try {
 		return withBackCompat(
```
`withBackCompat` might not be needed after your other changes to port parsing 🤔

ObliviousHarmony replied:
It's not, and, the conversion is covered by other tests.

noahtallen reviewed (commented):
Alright, made it through the non test files 😅 

Overall it definitely looks more clear and less clever. :P I did leave a note in one or two places where I think there could be more indirection than there needs to be

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/env-config-parsing

ObliviousHarmony committed:
Clarified Variable

ObliviousHarmony committed:
Fixed error message

ObliviousHarmony committed:
Removed Unnecessary `testsPort` Parsing

Since we're already handling this in the new parser, there's
no need to perform this translation here.

noahtallen commented on packages/env/test/parse-config.js:
```diff
@@ -1,61 +1,102 @@
+
+		const config = await loadConfig( path.resolve( '/test/gutenberg' ) );
+
+		expect( config ).toMatchSnapshot();
```
I don't dislike having some snapshot tests here, but I think it would be valuable to make a couple of assertions on the final data too. (For example, in this case, making an assertion on the core source and port)

I think it's easy for snapshot tests to be ignored, and it's harder to tell what the test is trying to verify. I think before I had just added one as a catch-all

ObliviousHarmony replied:
I added another integration test and a some assertions.

noahtallen commented on packages/env/test/parse-config.js:
we should move this to config/test -- not sure why it was in the root test directory before

noahtallen reviewed (commented):
Note for follow-up: we should probably add some documentation for the config parsing at some point

noahtallen commented:
Also, you make a lot of good points above :) My subjective take is that some of this makes the code more verbose than it really needs to be. Which isn't a huge problem, but I think there are one or two layers of indirection that are a bit difficult to grok.

That said, I think when we get some documentation in place (like "how to add a config key" or "how to add an env var override"), it really won't be a problem

ObliviousHarmony committed:
Merge branch 'trunk' into refactor/env-config-parsing

ObliviousHarmony committed:
Fixed Root `testsPort` Parsing

ObliviousHarmony committed:
Adjusted Config Type Definitions

These types will be more descriptive of their purpose.

ObliviousHarmony commented:
Thanks for the review @noahtallen!

I'm actually pretty happy about the above test failure because it shows that the test suite is really robust. When I removed the `testsPort` handling from `readRawConfigFile()` it caused the option to be silently discarded. This bug was then caught by the test suite.

I also aded an integration test for the environment variable overrides so that the tests cover all four configuration states.

noahtallen commented:
The Puppeteer 3 suite seems to be failing across all commits right now, unfortunately.

noahtallen reviewed (approved):
Nice, I'm happy with this! Thanks for the improvement.

talldan commented on packages/env/lib/config/test/parse-config.js:
```diff
@@ -0,0 +1,336 @@
+		expect( parsed ).toEqual( {
+			...DEFAULT_CONFIG,
+			coreSource: {
+				basename: 'gutenberg',
```
@ObliviousHarmony @noahtallen One side-effect of hard-coding the basename (and also testPath) like this is is that if someone checks out the gutenberg repo into a differently named folder, these tests will fail.

This is currently causing a few issues for CI jobs that run on security patches, as those are created on a differently named private repo.

I was wondering what your thoughts on a good fix would be?

A naive option could be `expect.any( String )`, but I think maybe this should all be mocked rather than being based off of the real file system.

ObliviousHarmony replied:
That's an interesting case @talldan. It seems like the best source would be to just generally avoid any relative paths and use fake absolute paths in our tests. [I've gone ahead and created an issue for this](https://github.com/WordPress/gutenberg/issues/50800).

talldan replied:
Thanks!
